### ISSUE TITLE
feat(voice): port Parakeet-TDT-0.6B-v2 to candle for streaming ASR

### ISSUE THREAD

**Issue by newhoggy** [rust-works/omni-voice#23]
> Replicated from rust-works/omni-dev#898 (https://github.com/rust-works/omni-dev/issues/898)

## Goal

Port [`mlx-community/parakeet-tdt-0.6b-v2`](https://huggingface.co/mlx-community/parakeet-tdt-0.6b-v2) (NVIDIA Parakeet-TDT 0.6B v2 — FastConformer encoder + TDT decoder + joiner, 600 M params, English-only) to pure Rust against the `candle 0.10.x` runtime already in tree per [ADR-0033](docs/adrs/adr-0033.md). Provide a `Transcriber` and `StreamingTranscriber` backend so [rust-works/omni-dev#806](https://github.com/rust-works/omni-dev/issues/806) can ship streaming ASR with substantially better WER than the existing `candle + whisper-tiny.en + LocalAgreement-2` path.

## Why this is a feature task, not a spike

Spike [rust-works/omni-dev#871](https://github.com/rust-works/omni-dev/issues/871) closed in 2026-05-23 with a **GO** recommendation. All four feasibility probes cleared:

1. **Op coverage** — zero blocking ops; three small custom helpers (rel-shift indexing, GLU split, STFT windowing) plus a conv-axis permute at load. No CUDA-like kernels needed.
2. **Weight conversion** — 697 fp32 tensors, 620 identity + 77 conv-axis permutations, MLX checkpoint names already match the natural candle module tree.
3. **Numerical parity (block-0 FFN sub-path)** — ~1e-3 relative error per sub-block, BLAS-level matmul drift between MLX/Accelerate and candle's `gemm`. Well inside the ±2 % WER bar.
4. **Encoder CPU RTF** — synthetic 24-block FastConformer encoder hits RTF 0.273 single-thread / 0.072 multi-thread on M1 Max for 5-min audio. 5.5× margin below the spike's abort gate.

Effort estimate from the spike: **~13 days P50, ~27 days P90 for one engineer**.

## Background — what's on the table

[rust-works/omni-dev#856](https://github.com/rust-works/omni-dev/issues/856) measured Parakeet at **3.65 % WER** on the committed 5-min Sherlock Holmes fixture. [rust-works/omni-dev#826](https://github.com/rust-works/omni-dev/issues/826) measured `candle + whisper-tiny.en + LocalAgreement-2` at **7.13 % WER** on the same fixture — and failed the RTF gate at 1.30 (LA-2 sliding-window re-inference dominated). The whisper path can't both clear RTF and deliver acceptable WER without a model upgrade; Parakeet's streaming-native TDT decoder eliminates the merger entirely and brings WER down ~3.5 pp.

Cross-runtime context from [`COMPARISON.md`](https://github.com/rust-works/omni-dev/blob/issue-871-parakeet-rust-port-spike/COMPARISON.md) on the rust-works/omni-dev#871 spike branch:

| Metric | parakeet (this port) | sherpa-onnx (rust-works/omni-dev#859) | candle + whisper-tiny.en (rust-works/omni-dev#826) |
|---|---|---|---|
| WER on the 5-min fixture | **3.65 %** | 6.81 % | 7.13 % |
| C++-free per ADR-0033 | ✅ | ❌ | ✅ |
| Streaming-native | ✅ (TDT decoder) | ✅ | ❌ (needs LA-2 merger) |

## Scope

**In:**

- **FastConformer encoder** (24 layers, `d_model=1024`, `n_heads=8`, `self_attention_model: rel_pos`, `att_context_size: [-1, -1]`, `conv_kernel_size=9`, `subsampling: dw_striding`). Implement against `candle-nn` primitives. Three custom helpers: relative-shift trick (~10 lines), GLU split (~3 lines), STFT windowing (~10 lines or pre-computed hop indices).
- **TDT (Token-and-Duration Transducer) decoder + joiner.** Predictor: 2-layer LSTM (`candle_nn::LSTM`, 1025-token embedding incl. blank). Joiner: 3× Linear + ReLU producing 1030 logits (1024 vocab + 1 blank + 5 durations).
- **Weight converter polished into production.** Promote `spike-parakeet-rust/convert_weights.py` to a stable, documented script under `scripts/` (or fold into `voice install-model` as a download-time conversion if the cache invariant supports it).
- **Tokenizer integration** — 1024-token BPE SentencePiece model. Reuse the existing `tokenizers` crate dep.
- **Streaming wrapper** implementing `StreamingTranscriber` + `AsyncAudioInput` from [rust-works/omni-dev#806](https://github.com/rust-works/omni-dev/issues/806). Per-chunk TDT decoder advance; emit `Partial` after each chunk's joint-net output, `Final` on TDT blank-runs that exceed a threshold, `Endpoint::StreamEnd` at input EOS.
- **Batch wrapper** implementing `Transcriber` (sync trait from rust-works/omni-dev#801). Same model, no chunked streaming.
- **`voice install-model` integration.** Knows how to fetch `mlx-community/parakeet-tdt-0.6b-v2` (2.47 GB) and run conversion. Records CC-BY-4.0 attribution in the model's local metadata.
- **Snapshot test fixtures** — extend the 5-min Sherlock Holmes WAV fixture from rust-works/omni-dev#826 with expected Parakeet output for the new backend. The committed `expected.txt` was generated against whisper-tiny.en; Parakeet's transcript will differ in detail (better proper-noun handling, different boundary recognitions) and needs its own `expected.parakeet.txt`.

**Out (deferred to follow-up issues):**

- **int8 quantisation** of weights or activations.
- **Multi-thread BLAS tuning** beyond candle's default rayon usage.
- **Multilingual Parakeet variants.** This port targets the English-only `tdt-0.6b-v2`; multilingual is additive.
- **CUDA / Metal backend support.** CPU only for v1; the candle runtime supports adding GPU later without changing the model code.
- **Promotion of the spike's `SPIKE.md` to ADR-0036.** That's a separate PR pending owner sign-off; this issue does the engineering.

## Acceptance criteria

The port lands when **all** of the following hold:

- [ ] `cargo build --release` succeeds with the new backend wired into `create_default_transcriber()` / `create_default_streaming_transcriber()`.
- [ ] **Token-level greedy decode parity:** on the committed 5-min Sherlock Holmes fixture, the candle Parakeet backend's transcript matches the [`newhoggy/parakeet-mlx@32b8034`](https://github.com/newhoggy/parakeet-mlx/commit/32b8034) fork's transcript to within **±2 % WER**. (Acknowledges bfloat16/float32 path differences and cross-runtime numerical drift may cause occasional token disagreement — see [rust-works/omni-dev#871's numerical-parity bar](https://github.com/rust-works/omni-dev/issues/871#numerical-parity-bar).)
- [ ] **Streaming-mode determinism:** repeated runs on the same fixture produce byte-equal event-log JSONL output. The event-log schema follows the `baseline/src/events.rs` shape from rust-works/omni-dev#826's spike (`Partial`/`Final`/`Endpoint`/`StreamEnd` with `wall_ms` + `audio_ms` annotations).
- [ ] **Insta snapshot tests** committed for: (a) the batch transcript on the 5-min fixture, (b) the streaming `Final`-only transcript on the same fixture, (c) a representative `Partial`-event sequence on a 30 s slice.
- [ ] **Boundary-artifact test** committed: streaming output on the full 5-min fixture vs. batch output on the same audio chopped into stitched 30 s clips. Diff recorded as an informational artefact (the streaming result is what we ship; the comparison documents the boundary errors that stitched-batch hides).
- [ ] **`voice install-model parakeet-tdt-0.6b-v2`** works end-to-end: download → convert → load. Resulting metadata file records CC-BY-4.0 attribution.
- [ ] **No production code changes outside `src/voice/**`** and `src/cli/voice/**`. Trait surface in `src/voice/transcriber.rs` extended with `StreamingTranscriber` and `AsyncAudioInput` (this is the rust-works/omni-dev#806 work that must land first or as part of this PR).
- [ ] **Clippy clean** (`cargo clippy -- -D warnings`); `cargo fmt` applied.
- [ ] **CHANGELOG entry** under the `[Unreleased]` section per the `changelog` skill conventions.

**Explicitly NOT acceptance gates:**

- CPU RTF on the production-target hardware. Per the user direction, **measurement happens after the port** (see Validation below). RTF/latency/RSS will be reported but won't block the merge.
- Performance parity with sherpa-onnx's 29 ms first-Partial p95. This port aims for *correct* streaming on candle CPU; optimisation lands as separate issues.

## Validation (post-port, not gating)

After the port merges, run the standard measurement battery from `baseline/` against the committed 5-min fixture:

| Metric | What | Action if outside budget |
|---|---|---|
| WER (%) | Against `expected.parakeet.txt` ground truth | File a fidelity issue; investigate weight conversion + numerical parity at deeper layers |
| RTF (wall / audio) | At 100 ms and 500 ms streaming cadence, single-thread CPU | If RTF > 0.5 at 500 ms on M1 Max: file int8 quantisation + multi-thread BLAS follow-ups |
| First-Partial p95 (ms) | Latency from chunk-end to `Partial` event | If > 1000 ms: investigate per-chunk overhead |
| RSS peak (MB) | RES during streaming run | If > 1 GB: investigate state retention (encoder KV cache + decoder LSTM state) |
| Determinism | Byte-equal event-log JSONL across 3 reruns | If non-deterministic: hard regression, investigate before any further measurement |

Report numbers in a follow-up PR comment, not as part of the port PR itself. If RTF > 0.5 on M1 Max baseline, file follow-up issues for int8 + multi-thread BLAS optimisation — but **don't gate the port commit on these**. Worst-case fallback if optimisation can't recover RTF: sherpa-rs + a single-paragraph amendment to [ADR-0033](docs/adrs/adr-0033.md) waiving C++-freeness for the streaming ASR backend.

## Implementation specifics

**Module layout** (additions under `src/voice/backends/`):

```
src/voice/backends/
  parakeet/
    mod.rs                    # CandleParakeetTranscriber, public surface
    encoder.rs                # FastConformer encoder (24-layer Conformer block stack)
    conformer_block.rs        # FFN + multi-head attention + conv module + FFN + LN
    attention.rs              # rel-pos multi-head attention (Shaw-style + Transformer-XL rel-pos)
    conv_module.rs            # GLU split + pointwise + depthwise + BN + SiLU + pointwise
    decoder.rs                # TDT predictor (LSTM) + joiner
    tokenizer.rs              # SentencePiece BPE wrapper
    audio.rs                  # mel-spectrogram via rustfft (already a root dep)
    streaming.rs              # StreamingTranscriber impl; per-chunk encoder advance + TDT loop
    weights.rs                # safetensors load via VarBuilder; conv-axis permute helpers
```

**Reference implementations** to mirror semantics against:

- **Patched MLX fork:** [`newhoggy/parakeet-mlx@32b8034`](https://github.com/newhoggy/parakeet-mlx/commit/32b8034) — ground truth for streaming behaviour. The running-stats normalisation fix in `StreamingParakeet` is load-bearing; the Rust streaming wrapper must replicate it.
- **Upstream:** [`senstella/parakeet-mlx`](https://github.com/senstella/parakeet-mlx) — non-streaming reference.

**Weight converter:** start from [`spike-parakeet-rust/convert_weights.py`](https://github.com/rust-works/omni-dev/blob/issue-871-parakeet-rust-port-spike/spike-parakeet-rust/convert_weights.py) on the rust-works/omni-dev#871 spike branch (108 lines). Polish it into a stable script under `scripts/convert_parakeet_weights.py` or fold conversion into `voice install-model` if the cache layout supports a one-shot transform at download time.

**Mel-spectrogram:** use `rustfft` (root dep at [Cargo.toml line 63](Cargo.toml#L63)). The `rfft` not being in candle is a documented partial; manual unfold-loop or pre-computed hop-index tables handle STFT windowing. If `src/voice/backends/candle.rs`'s mel path is compatible with Parakeet's frontend config (80 mel bins, `hop_length=160`, `win_length=400` at 16 kHz), reuse it; otherwise add a Parakeet-specific path.

**Trait integration.** Backend implements both:

- `Transcriber::transcribe(audio: Box<dyn AudioInput>) -> Result<Box<dyn EventStream>>` (sync batch, existing trait from rust-works/omni-dev#801)
- `StreamingTranscriber::transcribe_stream(audio: Box<dyn AsyncAudioInput>) -> Pin<Box<dyn Stream<Item = Result<TranscriptEvent>> + Send>>` (new from rust-works/omni-dev#806)

The shared encoder + tokenizer state can live on the backend type; streaming and batch differ only in how they drive the decoder loop. Wrap mutable per-call state (TDT LSTM hidden, encoder KV cache if any) in `Mutex` — same pattern as `src/voice/backends/candle.rs`.

**Endpoint detection.** TDT decoder emits blanks during silent regions; a run of N consecutive blank emissions (configurable, default ~1.5 s of silence at the encoder's frame rate) commits a `Final` and emits `Endpoint::SilenceGap`. No external VAD required for v1 — the model's own blank emissions are the endpoint signal. (Silero VAD as a pre-filter to skip encoder work on silent chunks is a future optimisation, deferred.)

**Tokenizer.** Parakeet ships its own SentencePiece BPE vocab (1024 tokens + blank). Convert the `.model` file to a tokenizers-crate-compatible format at download time, or load it directly via `tokenizers::Tokenizer::from_file` if the format is already compatible.

## Dependencies

- **Hard blocker:** [rust-works/omni-dev#806](https://github.com/rust-works/omni-dev/issues/806) — `StreamingTranscriber` + `AsyncAudioInput` trait surface must exist before this port can implement it. If rust-works/omni-dev#806's scope is shrunk to just the trait extension (per its post-ADR-0033 rewrite), that lands first; this port consumes it.
- **Soft dependency:** [rust-works/omni-dev#862](https://github.com/rust-works/omni-dev/issues/862) — landing the 5-min monologue fixture into `main`. Without it, this issue cherry-picks the three fixture files from `issue-826-streaming-asr-spike` directly.
- **Informed by:** [rust-works/omni-dev#871](https://github.com/rust-works/omni-dev/issues/871) — feasibility spike; the answer to "is this doable" lives in [`SPIKE.md`](https://github.com/rust-works/omni-dev/blob/issue-871-parakeet-rust-port-spike/SPIKE.md) on that branch. Lift the converter, the weight-name mapping, and the encoder-bench harness.
- **Informed by:** [rust-works/omni-dev#856](https://github.com/rust-works/omni-dev/issues/856) — quality baseline (3.65 % WER), streaming-fix fork pinned at `32b8034`.
- **Informed by:** [rust-works/omni-dev#826](https://github.com/rust-works/omni-dev/issues/826) — measurement harness (`baseline/{events.rs,wer.rs,idle.rs}`) and the 5-min fixture provenance.
- **Touches:** [ADR-0033](docs/adrs/adr-0033.md) — this port is the implementation of the candle commitment. If the fallback scenario (sherpa-rs waiver) is needed post-validation, that's a separate ADR amendment.

## Don't

- **Don't optimise pre-emptively.** Land the correct port first, measure, then optimise where measurement points. Specifically: no int8 quantisation, no manual SIMD, no multi-thread BLAS tweaks before the post-port measurement battery says they're needed.
- **Don't change the `StreamingTranscriber` trait shape mid-port** — that's the contract [rust-works/omni-dev#807](https://github.com/rust-works/omni-dev/issues/807) (`voice listen`) binds against.
- **Don't merge the spike code from rust-works/omni-dev#871's branch directly.** The spike's `spike-parakeet-rust/` crate is throwaway; the production port lives under `src/voice/backends/parakeet/` and is structured for long-term maintenance.
- **Don't ship without the CC-BY-4.0 attribution.** `voice install-model` must record the licence in the model's metadata file; document it in `docs/voice.md` (or wherever the model registry lives).
- **Don't expand scope to multilingual Parakeet variants** in this PR. That's a follow-up issue.
- **Don't gate the merge on CPU RTF.** Per the issue brief, measurement happens after.

## Context

- Parent product issue: [rust-works/omni-dev#806](https://github.com/rust-works/omni-dev/issues/806) — streaming ASR implementation
- Umbrella: [rust-works/omni-dev#799](https://github.com/rust-works/omni-dev/issues/799) — voice assistant build order
- Feasibility spike: [rust-works/omni-dev#871](https://github.com/rust-works/omni-dev/issues/871) — GO recommendation, ~13d P50 / ~27d P90 estimate
- Quality baseline: [rust-works/omni-dev#856](https://github.com/rust-works/omni-dev/issues/856) — 3.65 % WER reference, streaming-fix fork at [`newhoggy/parakeet-mlx@32b8034`](https://github.com/newhoggy/parakeet-mlx/commit/32b8034)
- Runtime-selection spike: [rust-works/omni-dev#826](https://github.com/rust-works/omni-dev/issues/826) — measurement harness + 5-min fixture origin
- Sibling baselines: [rust-works/omni-dev#858](https://github.com/rust-works/omni-dev/issues/858) (whisper-cpp), [rust-works/omni-dev#859](https://github.com/rust-works/omni-dev/issues/859) (sherpa-onnx), [rust-works/omni-dev#873](https://github.com/rust-works/omni-dev/issues/873) (moonshine-mlx)
- ADRs: [ADR-0033](docs/adrs/adr-0033.md) (candle for ASR), [ADR-0034](docs/adrs/adr-0034.md) (tract-onnx for speaker)
- Model card: [`mlx-community/parakeet-tdt-0.6b-v2`](https://huggingface.co/mlx-community/parakeet-tdt-0.6b-v2) — 600 M params, English-only, CC-BY-4.0
